storage.go

```diff
@@ -0,0 +1,164 @@
+
+// NewRedisStorage builds a new RedisStorage.
+func NewRedisStorage(c RedisConfig) (*RedisStorage, error) {
+	rdb := redis.NewClient(&redis.Options{
+		Addr:     fmt.Sprint(c.Host, ":", c.Port),
+		Username: c.Usr,
+		Password: c.Pwd,
+		DB:       c.DB,
+	})
+
+	if c.TTL == 0 {
+		c.TTL = defTTL
+	}
+
+	storage := &RedisStorage{
+		rdb:   rdb,
+		cache: cache{},
+		ttl:   time.Duration(c.TTL) * time.Hour,
+	}
+
+	var err error
+	storage.cache, err = storage.getRemoteChecks(context.Background())
+	if err != nil {
+		return nil, fmt.Errorf("Err retrieving checks: %v", err)
+	}
+
+	go storage.refresh()
+
+	return storage, nil
+}
+
+// GetAbortedChecks returns the list of UUIDs for the currently aborted checks.
+func (r *RedisStorage) GetAbortedChecks(ctx context.Context) ([]string, error) {
+	r.RLock()
+	defer r.RUnlock()
+
+	// Because we have mantained local cache in sync
+	// with remote storage, we can return local copy
+	// directly instead of performing requests to redis.
+	return r.cache, nil
+}
+
+// AddAbortedChecks adds the given checks to the current aborted checks list.
+func (r *RedisStorage) AddAbortedChecks(ctx context.Context, checks []string) error {
+	r.Lock()
+	defer r.Unlock()
+
+	// Because we have mantained local cache in sync
+	// with remote storage, we can add new checks to
+	// local cache and set that value in redis instead
+	// of performing extra requests to retrieve all
+	// remote values.
+	for i, c := range checks {
+		key := fmt.Sprint(checksKeyPrefix, c)
+		err := r.rdb.Set(ctx, key, c, r.ttl).Err()
+		if err != nil {
+			return fmt.Errorf("%d checks could not be aborted: %v", len(checks)-i, err)
+		}
+		r.cache = append(r.cache, c)
+	}
+
+	return nil
+}
+
+// getRemoteChecks retrieves the checks stored in redis.
+// Because we want to set a TTL per check, we had to store one entry per check,
+// so to retrieve them we have to perform a SCAN matching 'check:*' and retrieve
+// the value per each obtained key.
+// SCAN is performed in chunks so we don't block redis when there are many entries.
+func (r *RedisStorage) getRemoteChecks(ctx context.Context) ([]string, error) {
+	r.RLock()
+	defer r.RUnlock()
+
+	var (
+		err    error
+		cursor uint64
+		checks []string
+	)
+
+	match := fmt.Sprint(checksKeyPrefix, "*")
+	for {
+		var keys []string
+		keys, cursor, err = r.rdb.Scan(ctx, cursor, match, defScanChunk).Result()
+		if err != nil {
+			return nil, err
+		}
+		for _, k := range keys {
+			checkID, err := r.rdb.Get(ctx, k).Result()
+			if err != nil {
+				return nil, err
+			}
+			checks = append(checks, checkID)
+		}
+		if cursor == 0 {
+			break
+		}
+	}
+
+	return checks, nil
+}
+
+// refresh refreshes the redis storage local cache
+// periodically so checks that have been expired
+// remotely due to TTL, are also removed locally.
+func (r *RedisStorage) refresh() {
+	ctx := context.Background()
+	for {
+		time.Sleep(time.Duration(rfshPeriod) * time.Hour)
+		r.Lock()
+		r.cache, _ = r.getRemoteChecks(ctx)
```

Comment:
I would get the possible error here and log it, so we could now if there is any problem with Redis.

Reply:
This I thought about it, but because we are not losing information here, because in the end the refresh is only to get ride of checks that might have been expired from redis due to TTL, but we would not be losing checks information, and if the stream restarts and can't reach redis then it will fail to start and we would know, so I decided to not do it so we don't have to pass the logger to storage, but we can do it for sure, on it!

Reply:
Done @manelmontilla , I also took the chance to refactor a little bit the storage layer so we can mock remote storage connector and implement easier testing for this part. Will try to address that later on also.